### PR TITLE
unreachable_from_main: Do not consider type annotations when computing reachability

### PR DESCRIPTION
**Pull request by srawlins**
# Description

A reference in a type annotation is not good enough to count as "reachable". Marking a type as reachable only because it was seen in a type annotation would be a miscategorization if the type is never instantiated or subtyped.

This was inspired by a comment @bwilkerson made in an earlier review.

CC @goderbauer This is the last change I plan for this rule 😁

### CODE REVIEW

**Review by bwilkerson — APPROVED** [review]
I don't remember the comment, but it sounds reasonable that type annotations don't count. It would be interesting to know whether there are any new violations, especially false positives, as a result of this change.

**Comment by srawlins** [comment]
Yeah I'll try and get some numbers.

**Comment by srawlins** [comment]
It's good you asked and it's good I tested on a few packages! I discovered that this was not sound in the case of type aliases: a type alias may be used in any type annotation, and that is enough to consider it "used." A function type or a record type, in particular, are not instantiated by referring to the alias by name (and interface types don't need to be either).

This PR is quite different now; another review would be appreciated.

File: test/rules/unreachable_from_main_test.dart [hunk 1]
```diff
@@ -1007,6 +1009,109 @@ int x = 1;
     ]);
   }
 
+  test_typedef_reachable_eferencedInObjectPattern() async {
```
Comment (bwilkerson) [inline test/rules/unreachable_from_main_test.dart:1012]: nit: typo in "eferenced"

Comment (srawlins) [reply test/rules/unreachable_from_main_test.dart:1012]: Thanks! Fixed

**Pull request merged**